bgpd: SRV6 L3VPN + RouteReflector topology modifications

rvtch opened the pull request:
1. Processing of prefixes (BGP_UPDATE) on RR node is set to the following algorithm (accordingly to RFC specs):
A - AS local = AS of peer
R - peer configured as RR Client
L - prefix is locally originated
S - send BGP update
D - drop BGP update

A R L Action
============
```
- - -  S
- - +  S
- + -  S
- + +  S
+ - -  D
+ - +  S
+ + -  S
+ + +  S
```

2. Minor changes accordingly to maintainers' proposals in #11348

3. topotests for processing SRV6 L3VPN case by RouteReflector

4. Hardcoded Locator Block Length, Locator Node Length, Transposition Length, Transposition Offset in SRv6 SID Sub-Sub-TLV. Now they are correctly calculated from the locator mask (/48 or/64) in the configuration for compatibility with Cisco IOS-XR.

5. A length of SRv6 SID value length of SRv6 SID Sub-TLV is aligned to requirements of p.4 of draft-ietf-bess-srv6-services-15 for L3 VPN services.

6. SRv6 Endpoint Behavior field of SRv6 SID Sub-TLV is set accordingly to IANA SRv6 Endpoint Behaviors asignments.

7. MP_REACH_NLRI Next Hop is now IPv6 for both IPv4 and IPv6 prefixes.

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

Commit:
bgpd: SRv6 L3 VPN bugs and drawbacks

1. Hardcoded Locator Block Length, Locator Node Length, Transposition Length, Transposition Offset in SRv6 SID Sub-Sub-TLV. Now they are correctly calculated from the locator mask (/48 or/64) in the configuration for compatibility with Cisco IOS-XR.
2. A length of SRv6 SID value length of SRv6 SID Sub-TLV is aligned to requirements of p.4 of draft-ietf-bess-srv6-services-15 for L3 VPN services.
3. SRv6 Endpoint Behavior field of SRv6 SID Sub-TLV is set accordingly to IANA SRv6 Endpoint Behaviors asignments.
4. MP_REACH_NLRI Next Hop is now IPv6 for both IPv4 and IPv6 prefixes.
5. Improved L3 VPN scalability from 128 to 64K in label stack calculation.

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

Commit:
bgpd: SRV3 L3VPN locator+transposition len bug fix

SRV6 L3VPN - locator block len, locator node len, transposition offset,
transposition len check (get from configuration/get from BGP_UPDATE
attributes)

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

Commit:
bgpd: SRV6 L3VPN endpoint at BGP_UPDATE bug fix

Add IANA endpoint behavior at processing BGP_UPDATE

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

rvtch committed:
bgpd: processing endpoint behavior at BGP_UPDATE bug fix

SRV6 L3VPN - select correct endpoint behavior at building attributest for BGP_UPDATE message

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

celotajs committed:
Merge branch 'FRRouting:master' into jovvo-master-srv6-check

rvtch committed:
tests: adapt test bgp_srv6l3vpn_to_bgp_vrf for new calc label algorithm

Adapt test bgp_srv6l3vpn_to_bgp_vrf for new calc label algorithm

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

rvtch committed:
bgpd: code style corrections

Code style corrections

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

rvtch committed:
bgpd: SRV6 L3VPN RouteReflector modifications

Modifications:
 - build BGP_UPDATE attr for SRV6 L3VPN - RouteReflector case
 - implementation of SRV6 L3VPN prefix announce algorithm:
    A - AS local = AS of peer
    R - peer configured as RR Client
    L - prefix is locally originated
    S - send BGP update
    D - drop BGP update

    A R L Action
    ============
    - - -  S
    - - +  S
    - + -  S
    - + +  S
    + - -  D
    + - +  S
    + + -  S
    + + +  S

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

rvtch committed:
bgpd: ipv4 processing for srv6 l3vpn bug fix

Set correct build attribute addresses (global|global for nexthop self) for srv6 l3vpn case

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

rvtch committed:
bgpd: !strcmp()->strmatch() change as suggested by ton31337 in #11348 pull request

!strcmp()->strmatch() change as suggested by ton31337 in #11348 pull request

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

rvtch committed:
bgpd: code style corrections

Code style corrections

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

ton31337 commented:
Dup https://github.com/FRRouting/frr/pull/11348?

rvtch commented:
@[ton31337](https://github.com/ton31337) may be better way is to close this PR without merge and I'll merge changes into https://github.com/FRRouting/frr/pull/11348?

louberger commented:
this is changing rr updae route selection...

louberger reviewed (changes requested):
impact to route selection is failing labnci

rvtch commented:
@[louberger](https://github.com/louberger) Could you provide please with the details of error.

rvtch committed:
Merge branch 'master' into jovvo-srv6-check-2

rvtch committed:
bgpd: added tests (Route Reflector); bug fix (func subgroup_process_announce_selected for srv6 l3vpn route announcement)

added tests (Route Reflector); bug fix (func subgroup_process_announce_selected for srv6 l3vpn route announcement)

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

rvtch committed:
Merge branch 'jovvo-srv6-check-2' of github.com:jovvo-net/frr into jovvo-srv6-check-2

rvtch committed:
bgpd: SRV6L3VPN RouteReflector topotest fix

RouteReflector topotest fix

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

rvtch committed:
bgpd: SRV6L3VPN RouteReflector topotest fix

RouteReflector topotest fix (set kernel version for RR kernel test to 5.15)

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

rvtch committed:
Merge branch 'master' into jovvo-srv6-check-2

rvtch reviewed (commented):
Fixed (added topotests for RR case).

rvtch reviewed (commented):
Fixed

rvtch commented:
@ton31337 @louberger could you kindly review this pr to unlock further srv6 implementation

rvtch committed:
Merge branch 'master' into jovvo-srv6-check-2

rvtch committed:
tests: adapt test bgp_srv6l3vpn_to_bgp_vrf for modified calc label algorithm + code style corrections

Adapt test bgp_srv6l3vpn_to_bgp_vrf for new calc label algorithm

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

celotajs commented:
Hi All,
Dear @louberger and @louberger! Could you perform the review for this PR please? We have another set of SRv6 code to submit, but this hanged PR locks everything. Thank you very much in advance.

rvtch committed:
Merge branch 'master' into jovvo-srv6-check-2

rvtch reviewed (commented):
Checked

proelbtn commented on bgpd/bgp_mplsvpn.c:
```diff
@@ -1562,34 +1563,63 @@ void vpn_leak_from_vrf_update(struct bgp *to_bgp,	     /* to */
+				: (afi == AFI_IP6
+					   ? IANA_SEG6_ENDPOINT_BEHAVIOR_END_DT6
+					   : 0xffff);
+		if (chunk) {
+			static_attr.srv6_l3vpn->loc_block_len =
+				chunk->block_bits_length;
+			static_attr.srv6_l3vpn->loc_node_len =
+				chunk->node_bits_length;
+			static_attr.srv6_l3vpn->transposition_len =
+				chunk->function_bits_length ?: 16;
+			static_attr.srv6_l3vpn->transposition_offset =
+				chunk->block_bits_length +
+				chunk->node_bits_length;
+		} else {
+			static_attr.srv6_l3vpn->loc_block_len =
+				from_bgp->vpn_policy[afi]
+					.tovpn_sid_locator->block_bits_length;
+			static_attr.srv6_l3vpn->loc_node_len =
+				from_bgp->vpn_policy[afi]
+					.tovpn_sid_locator->node_bits_length;
+			static_attr.srv6_l3vpn->transposition_len =
+				from_bgp->vpn_policy[afi]
+					.tovpn_sid_locator
+					->function_bits_length;
+			static_attr.srv6_l3vpn->transposition_offset =
+				from_bgp->vpn_policy[afi]
+					.tovpn_sid_locator->block_bits_length +
+				from_bgp->vpn_policy[afi]
+					.tovpn_sid_locator->node_bits_length;
+			;
+		}
```
Why do you re-calculate these parameters?

The SRv6 locator from which SRv6 SID is allocated is saved in `from_bgp->vpn_policy[afi].tovpn_sid_locator`. So, you don't have to find locators and use them instead of saved ones.

And, current SRv6 locators have function length because user-specified or default function length is saved in srv6_locator structure. We don't have to consider the default function length.

### References

- https://github.com/FRRouting/frr/blob/master/bgpd/bgp_mplsvpn.c#L770-L791
- https://github.com/FRRouting/frr/blob/master/bgpd/bgp_mplsvpn.c#L679-L680
- https://github.com/FRRouting/frr/blob/master/zebra/zebra_srv6_vty.c#L274-L370

rvtch replied:
Thank you for your feedback. You are correct. I just submitted the update.

rvtch committed:
bgpd: SID for SRV6 selection update

SID for SRV6 selection update (proelbtn suggestion)

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

rvtch committed:
bgpd: code style corrections

Code style corrections

Signed-off-by: Roman Vartichan <roman.vartichan@gmail.com>

rvtch committed:
Merge branch 'master' into jovvo-srv6-check-2

riw777 commented:
@frrbot autoclose in 1 month
